File: app/src/main/java/com/chuongvd/app/signal/data/source/local/datasource/UserLocalDataSource.java
package com.chuongvd.app.signal.data.source.local.datasource;

import android.arch.lifecycle.LiveData;
import com.chuongvd.app.signal.AppExecutors;
import com.chuongvd.app.signal.data.database.AppDatabase;
import com.chuongvd.app.signal.data.entity.UserEntity;

public class UserLocalDataSource implements IUserLocalDataSource {
    private static UserLocalDataSource sInstance;

    private final AppDatabase mDatabase;
//    private AppExecutors mAppExecutors;

    private UserLocalDataSource(AppDatabase database) {
        mDatabase = database;
    }

    public static UserLocalDataSource getInstance(AppDatabase database) {
        if (sInstance == null) {
            synchronized (UserLocalDataSource.class) {
                if (sInstance == null) {
                    sInstance = new UserLocalDataSource(database);
                }
            }
        }
        return sInstance;
    }

    @Override
    public void saveUser(UserEntity user) {
        mDatabase.userDao().insertUser(user);
    }

    @Override
    public LiveData<UserEntity> getUser() {
        return mDatabase.userDao().loadUser();
    }

    @Override
    public String getAccessToken() {
        if (null == getUser().getValue()) return null;
        return getUser().getValue().getAccessToken();
    }
}

File: app/src/main/java/com/chuongvd/app/signal/data/source/local/datasource/UserLocalDataSource.kt
package com.chuongvd.app.signal.data.source.local.datasource

import android.arch.lifecycle.LiveData
import com.chuongvd.app.signal.data.database.AppDatabase
import com.chuongvd.app.signal.data.entity.UserEntity

class UserLocalDataSource
//    private AppExecutors mAppExecutors;

private constructor(private val mDatabase: AppDatabase) : IUserLocalDataSource {

  override val user: LiveData<UserEntity>
    get() = mDatabase.userDao().loadUser()

  override val accessToken: String?
    get() = if (null == user.value) null else user.value!!.accessToken

  override fun saveUser(user: UserEntity) {
    mDatabase.userDao().insertUser(user)
  }

  companion object {
    private var sInstance: UserLocalDataSource? = null

    fun getInstance(database: AppDatabase): UserLocalDataSource {
      if (sInstance == null) {
        synchronized(UserLocalDataSource::class.java) {
          if (sInstance == null) {
            sInstance = UserLocalDataSource(database)
          }
        }
      }
      return this.sInstance!!
    }
  }
}
